### ISSUE TITLE
Pen strokes disappear problem in "export as PDF".

### ISSUE THREAD

**Issue by Steven-Hsie** [xournalpp/xournalpp#4381]
(Please complete the following information, and then delete this line)

**Affects versions :**
 - OS: Linux Pop!_OS 22.04
 - Desktop environment: X11
 - Version of libgtk: 3.24.33
 - Version of Xournal++: 1.1.2
 - Installation method: ppa:apandada1/xournalpp-stable

**Describe the bug**
This bug only happened when using a stylus pen. In stroke recognizer mode, some strokes recognization will be failed. After this failure, any new drawing on the same page will not be shown in the exported PDF.

**To Reproduce**
1. Setting "Input System -> Input stabilization -> Averaging method" as "Velocity based Gaussian weight".
2. Turn on stroke recognizer mode.
3. Draw a " Ↄ " with stylus pen (in order to make the shape be recognized as a circle). Sometimes the shape just drawn will disappear.
4. Draw any new stroke on the same page.
5. Export as PDF.
6. Check the exported PDF file will discover that, any new drawing after disappearance on the same page just gone.

**Expected behavior**
The exported PDF should be completed.

**Screenshots of Problem**
![Screenshot from 2022-11-04 18-35-07](https://user-images.githubusercontent.com/104367376/199952698-3266f753-d8e4-41cd-9d1d-dd372f1f67b7.png)

**Additional context**
Add any other information or context about the problem here.

**Comment by rolandlo** [comment]
Thanks for reporting. Does that issue only appear if input stabilization (with "Velocity based Gaussian weight"s) is activated?

**Comment by Steven-Hsie** [comment]
According to the testing now, it is.

**Comment by rolandlo** [comment]
I couldn't reproduce the issue (neither with version 1.1.2 nor with the most current version 1.1.2+dev). The PDF export shows all strokes for me. What parameter value (sigma) do you have for the "Velocity based Gaussian weights"?

**Comment by Steven-Hsie** [comment]
Weird, I can reproduce this bug on two devices, Surface Go 2 and ENVY x360. The parameter value I have is 0.50. Notice that everything should be drawn with a stylus pen, and do Step3 until circle recognition fails (the circle just drawn disappear) before going to Step4.
![Screencast_2022-11-05_13_22_12_AdobeExpress](https://user-images.githubusercontent.com/104367376/200104096-5d5f6898-f0b2-453d-8899-0c0487fd1582.gif)

**Comment by bhennion** [comment]
Could you try running the app from terminal `$ xournalpp` and post the terminal output? Do you have any terminal output when the recognition fails?

**Comment by Steven-Hsie** [comment]
I didn't see any extra ouput when recognition fails. 

The terminal output of running app from terminal:
** Message: 18:37:36.481: TEXTDOMAINDIR = (null), Platform-specific locale dir = /usr/share/xournalpp/../locale, chosen directory = /usr/share/xournalpp/../locale
ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
** Message: 18:37:36.634: Plugin "MigrateFontSizes" UI initialized

**Comment by bhennion** [comment]
And upon exporting either?

**Comment by Steven-Hsie** [comment]
No, I didn't see any terminal output either. Or maybe it won't show up automatically, just because I don't know how to make it being shown?

**Comment by bhennion** [comment]
No no, something would appear without you doing anything... This is pretty weird.

**Comment by Steven-Hsie** [comment]
It is what I see.
![Screencast_2022-11-06_23_37_50_AdobeExpress](https://user-images.githubusercontent.com/104367376/200180217-d0047932-6e48-4513-868e-3e2f46f127b3.gif)

**Comment by bhennion** [comment]
And what happens if you save the file (as .xopp), and open it back again? Are the "phantom" strokes still there?

**Comment by Steven-Hsie** [comment]
Yes, they are still there, and also can not see them on the exported PDF.

**Comment by rolandlo** [comment]
Please post the `.xopp`-file then. Maybe we can find something unusual in it.

**Comment by Steven-Hsie** [comment]
Here is it.
[2022-11-07-Note-14-05.zip](https://github.com/xournalpp/xournalpp/files/9948568/2022-11-07-Note-14-05.zip)

**Comment by rolandlo** [comment]
The issue can be reproduced from the `.xopp`-file. There are `nan`'s in the pressure values, like here (at the end):
```xml
<stroke tool="pen" ts="0" fn="" color="#3333ccff" width="2.26000000 0.83237559 0.95219130 1.03831700 1.17056816 1.22386643 1.25370286 1.27947536 1.32089183 1.34962508 1.36939135 1.38563099 1.40934467 1.43060238 1.45121410 1.46785137 1.49094318 1.51167657 1.55160051 1.58033621 1.63383675 1.70196960 1.73992313 1.77455372 1.80937973 1.86317943 1.89012718 1.90879030 1.92533391 1.95377148 1.96711816 1.97853453 1.99114111 2.01226640 2.01915298 2.02533057 2.03143864 2.04195319 2.04895459 2.05619670 2.06341051 2.07407904 2.08451798 2.09556931 2.10578784 2.12101601 2.13618314 2.15310168 2.17000990 2.20335219 2.22701553 2.24684811 2.26718334 2.29791925 2.30850298 2.32018696 2.33144352 2.05413808 1.77683264 1.49952720 1.22222176 0.94491632 0.66761088 0.39030544 0.11300000 nan">
```
The file can't be opend with version 1.1.2+dev because of that. Would you mind trying out version 1.1.2+dev and see if you can generate a corrupted file like that with that version as well?

**Comment by rolandlo** [comment]
The `nan`' appear 29 times in the .xopp-file and they are either at the end of the pressure values or at the beginning (but only in cases where there are exactly 1 or 2 pressure values). 
Here is the repaired .xopp-file (I manually replaced all `nan`s with `0`-s and let version 1.1.2+dev run over it.
[2022-11-07-Note-14-05_repaired.zip](https://github.com/xournalpp/xournalpp/files/9948767/2022-11-07-Note-14-05_repaired.zip)

**Comment by Steven-Hsie** [comment]
In version 1.1.2+dev, it still has the recognization failure problem in Step3 (the circle just drawn disappeared). But the PDF export problem is fixed. Moreover, I found out that sometimes rectangles and lines also cause recognization failure.

**Comment by rolandlo** [comment]
@Steven-Hsie I'm trying to find the reason why you get `nan`s in the pressure values. Does the preprocessor method (inertia, deadzone, none) matter in the preferences? Does it matter if you activate the "Finalize the stroke" checkbox or not?

**Comment by rolandlo** [comment]
In @Steven-Hsie's file whenever the last pressure value is `nan` the last two stroke points have the same coordinates. I think the problem lies in the stroke finalizer. When the `distance` is computed in line 119 inside

https://github.com/xournalpp/xournalpp/blob/ef3a687fe5e94df17cb756d708935d4a920ef02d/src/core/control/tools/StrokeStabilizer.cpp#L105-L138

the scalar product and `normAB` or `normSquaredBC` can be zero (when either the last two points of the stroke are equal or the coordinates of the last event coincide with the coordinates of the last point of the stroke). In that case `distance` becomes `nan` and finally `coeff` and the pressure value `B.z` are set to `nan`.

The shape recognizer plays only the role of averaging pressure values. If the last one is `nan`, the computed `width` is `nan`.

@bhennion Should we fix the case where B is either A or C here or is it reasonable to assume that B is different from A and C (i.e. the problem lies at a different place)?
@Steven-Hsie Can you check if the issue is gone if you disable the "Finalize stroke" checkbox in the input system tab of the preferences?

**Comment by rolandlo** [comment]
@Steven-Hsie You could please try out the version from https://github.com/xournalpp/xournalpp/pull/4401#issuecomment-1312445238, check if the issue is still present or not, and report any unusual console messages?

You will probably have to use the .AppImage inside the Ubuntu 18.04 package when running Pop!_OS 22.04.

**Comment by bhennion** [comment]
I guess C could be anything, you're right. Let's see if your commit in #4401 fixes it then. Thanks for taking care of this!

**Comment by Steven-Hsie** [comment]
@rolandlo this issue is gone when I disable the "Finalize stroke"!

**Comment by Steven-Hsie** [comment]
The problem is fixed in the version from #4401! Thanks for making this app better and better.
There is a console message below, I don't know if it is unusual.

** Message: 20:06:58.930: TEXTDOMAINDIR = (null), Platform-specific locale dir = /usr/share/xournalpp/../locale, chosen directory = /usr/share/xournalpp/../locale
ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
** Message: 20:06:59.082: Plugin "MigrateFontSizes" UI initialized

**Comment by rolandlo** [comment]
Thanks for testing @Steven-Hsie ! I'm glad it worked. Yes, there is nothing unusual in the console messages.
